drape_frontend/gps_track_renderer.cpp

```diff
@@ -0,0 +1,304 @@
+#include "drape_frontend/gps_track_renderer.hpp"
+#include "drape_frontend/visual_params.hpp"
+
+#include "drape/glsl_func.hpp"
+#include "drape/shader_def.hpp"
+#include "drape/vertex_array_buffer.hpp"
+
+#include "indexer/scales.hpp"
+
+#include "base/logging.hpp"
+
+namespace df
+{
+
+namespace
+{
+
+int const kMinVisibleZoomLevel = 14;
+
+size_t const kAveragePointsCount = 512;
+
+float const kRadiusInPixel[] =
+{
+  // 1   2     3     4     5     6     7     8     9     10
+  1.0f, 1.0f, 1.0f, 1.0f, 1.0f, 1.0f, 1.0f, 1.0f, 1.0f, 1.0f,
+  //11   12    13    14    15    16    17    18    19     20
+  1.0f, 1.0f, 1.0f, 1.0f, 1.0f, 2.0f, 2.0f, 3.0f, 4.0f, 5.0f
+};
+
+double const kMinSpeed = 1.0;
+double const kAvgSpeed = 2.0;
+double const kMaxSpeed = 5.0;
+dp::Color const kMinSpeedColor = dp::Color(255, 0, 0, 255);
+dp::Color const kAvgSpeedColor = dp::Color(255, 255, 0, 255);
+dp::Color const kMaxSpeedColor = dp::Color(0, 255, 0, 255);
+
+bool GpsPointsSortPredicate(GpsTrackPoint const & pt1, GpsTrackPoint const & pt2)
+{
+  return pt1.m_id < pt2.m_id;
+}
+
+dp::Color InterpolateColors(dp::Color const & color1, dp::Color const & color2, double t)
+{
+  double const r = color1.GetRed() * (1.0 - t) + color2.GetRed() * t;
+  double const g = color1.GetGreen() * (1.0 - t) + color2.GetGreen() * t;
+  double const b = color1.GetBlue() * (1.0 - t) + color2.GetBlue() * t;
+  double const a = color1.GetAlfa() * (1.0 - t) + color2.GetAlfa() * t;
+  return dp::Color(r, g, b, a);
+}
+
+} // namespace
+
+GpsTrackRenderer::GpsTrackRenderer(TRenderDataRequestFn const & dataRequestFn)
+  : m_dataRequestFn(dataRequestFn)
+  , m_needUpdate(false)
+  , m_waitForRenderData(false)
+  , m_startSpeed(0.0)
+  , m_endSpeed(0.0)
+  , m_startColor(kMaxSpeedColor)
+{
+  ASSERT(m_dataRequestFn != nullptr, ());
+  m_points.reserve(kAveragePointsCount);
+  m_handlesCache.reserve(8);
+}
+
+float GpsTrackRenderer::CalculateRadius(ScreenBase const & screen) const
+{
+  double const zoomLevel = my::clamp(fabs(log(screen.GetScale()) / log(2.0)), 1.0, scales::UPPER_STYLE_SCALE + 1.0);
```

Comment:
log(2.0) can be calculated only once, it is not an constexpr function, so, it will be calculated every time.